### PR TITLE
Properly deal with absolute paths in unit tests, part 1

### PR DESCRIPTION
**Pull request by rmohr**
**What this PR does / why we need it**:

This is the first PR of a series which fixes a lot of unit tests which are not properly written. They had:

 * shared contexts with global variables and functions which leads to:
   * unwanted side effects if unit tests are run in parallel
   *  complicated BeforeEach and AfterEach logic
 * absolute paths hardcoded which  made it impossible to run them as non-root
 * tests for a lot of golbal functions without context which
   * made it very hard to mock them out
   * required deep preparations when testing higher-level modules

Now they do the following:

 * Introduce structs and  interfaces which
   * have their own state, where for instance absolute paths can be cleanly replaced with temporary directories
   * can easily be mocked out in higher level tests due to the usage of interfaces 

Not everything is absolutely perfect in this PR. Right now the main focus lies on requiring not root and being able to get rid of absolute paths.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/emptydisk/emptydisk.go [hunk 1]
```diff
@@ -10,22 +10,25 @@ import (
 	ephemeraldiskutils "kubevirt.io/kubevirt/pkg/ephemeral-disk-utils"
 )
 
-var EmptyDiskBaseDir = "/var/run/libvirt/empty-disks/"
+const emptyDiskBaseDir = "/var/run/libvirt/empty-disks/"
```
Comment (ezrasilvera) [inline pkg/emptydisk/emptydisk.go:14]: Do we want to take advantage of this PR and move the directory constants out of the individual go files into some centralize place ?  This might help us switching between root/nonroot ..
We can also do it on separate PR ..

Comment (rmohr) [reply pkg/emptydisk/emptydisk.go:14]: I agree with you that this should be done, but let's give it a dedicated PR. I would assume that I would now move it to A and a follow up PR just for this would again move it somewhere else.

**Review by EdDev — COMMENTED** [review]
I have tried to suggest a slightly different path in the first commit, which will probably change things in the following ones.
I hope I managed to express my thoughts clearly.

I generally found the use of interfaces that wrap the whole object methods as problematic. This usually comes with the definition of interfaces side by side with the concrete implementation. But interfaces are there to decouple between the caller and the concrete implementation.

In any case, the overall idea of decoupling and loosing the dependency is well worth this PR.

File: pkg/emptydisk/emptydisk.go (hunk 1 above)
Comment (ezrasilvera) [reply pkg/emptydisk/emptydisk.go:14]: @EdDev maybe "usually", but not in the case we are talking about. I was specifically talking about paths for volumes and disks. 
We have many different volume types,  sometimes implemented in different packages,  today the path is hardcoded in those files/packages. Just changing those to Const and leaving them local in those files won't help much. 
The base directory should be a global value (either through global variable or better a global function exposing the value) , that all other disk/volume related packages should use.  
The full path should be of type "baseDir" + "specificVolumePath"  where the first part is global and the second can be define on the particular package.

File: pkg/virt-handler/hotplug-disk/mount_test.go [hunk 2]
```diff
@@ -248,8 +250,11 @@ var _ = Describe("HotplugVolume block devices", func() {
 
 	It("mountBlockHotplugVolume and unmountBlockHotplugVolumes should make appropriate calls", func() {
 		blockSourcePodUID := types.UID("fghij")
-		hotplugdisk.SetKubeletPodsDirectory(tempDir)
-		targetPodPath := filepath.Join(tempDir, string(m.findVirtlauncherUID(vmi)), "volumes/kubernetes.io~empty-dir/hotplug-disks")
+		mknodCommand = func(deviceName string, major, minor int64, blockDevicePermissions string) ([]byte, error) {
```
Comment (stu-gott) [inline pkg/virt-handler/hotplug-disk/mount_test.go:253]: This function appears several times in this test module. Is there a way to reduce copy/paste?

Clever use of MkdirAll by the way.

File: pkg/emptydisk/emptydisk.go [hunk 3]
```diff
@@ -10,22 +10,25 @@ import (
...
 
-func CreateTemporaryDisks(vmi *v1.VirtualMachineInstance) error {
+type emptyDiskCreator struct {
+	emptyDiskBaseDir string
+	discCreateFunc   func(filePath string, size string) error
+}
```
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/emptydisk/emptydisk.go:18]: Why would I not want to know if my code does something like `vmi.Spec.Volumes[0]` instead of selecting all empty disks in a VMI?

File: pkg/hooks/manager.go [hunk 4]
```diff
@@ -50,18 +51,23 @@ var manager *Manager
...
 
 func GetManager() *Manager {
 	once.Do(func() {
-		manager = &Manager{CallbacksPerHookPoint: make(map[string][]*callBackClient)}
+		manager = getManager(HookSocketsSharedDirectory)
 	})
 	return manager
 }
 
+func getManager(baseDir string) *Manager {
```
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/hooks/manager.go:65]: It only gets a new manager on the first call. I only made it private.

[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/hooks/manager.go:65]: ah never mind.

File: pkg/hotplug-disk/hotplug-disk.go [hunk 5]
```diff
@@ -30,30 +30,55 @@ import (
...
 )
 
+const (
+	hotplugDisksKubeletVolumePath = "volumes/kubernetes.io~empty-dir/hotplug-disks"
+)
+
 var (
-	podsBaseDir = filepath.Join(util.HostRootMount, util.KubeletPodsDir)
+	// visible for testing
+	TargetPodBasePath = func(podBaseDir string, podUID types.UID) string {
```
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/hotplug-disk/hotplug-disk.go:39]: Yes it is used in other tests too, removing some duplication. It is not perfect but better.

File: pkg/virt-handler/hotplug-disk/mount_test.go [hunk 6]
```diff
@@ -996,8 +1007,8 @@ var _ = Describe("HotplugVolume volumes", func() {
 		Expect(bytes).To(Equal(expectedBytes))
 		_, err = os.Stat(fileSystemVolume)
 		Expect(err).ToNot(HaveOccurred())
-		_, err = os.Stat(blockVolume)
-		Expect(err).ToNot(HaveOccurred())
+		//		_, err = os.Stat(blockVolume)
+		//		Expect(err).ToNot(HaveOccurred())
```
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/virt-handler/hotplug-disk/mount_test.go:1011]: Yes.

File: pkg/hooks/manager.go [hunk 7]
```diff
@@ -97,7 +103,7 @@ func collectSideCarSockets(numberOfRequestedHookSidecars uint, timeout time.Dura
 					continue
 				}
 
-				callBackClient, notReady, err := processSideCarSocket(HookSocketsSharedDirectory + "/" + socket.Name())
+				callBackClient, notReady, err := processSideCarSocket(filepath.Join(m.hookSocketSharedDirectory + "/" + socket.Name()))
```
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/hooks/manager.go:106]: Yes.

**Comment by rmohr** [comment]
> I generally found the use of interfaces that wrap the whole object methods as problematic. This usually comes with the definition of interfaces side by side with the concrete implementation. But interfaces are there to decouple between the caller and the concrete implementation.

Yes, I prefer to wait with more focused interfaces which abstract more from the concrete implementation until it is needed. I don't see the need now. For testing I just want the concrete one replaced and for as long as there is no design where multiple of such interfaces would need a specific form to work together in a framework it is just more code and guessing on how we would want hypothetical cases work.

File: pkg/virt-handler/hotplug-disk/mount_test.go (hunk 2 above)
Comment (rmohr) [reply pkg/virt-handler/hotplug-disk/mount_test.go:253]: Most of them are slightly different. I am not sure if it is great usage. I think that needs a bigger rewrite later.

File: pkg/hotplug-disk/hotplug-disk.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/hotplug-disk/hotplug-disk.go:39]: You mean that the tests are assigning a stub instead of the original function?
I could not see that, maybe I missed something. I just saw it used with `hotplugdisk.TargetPodBasePath` which can work with a simple exported function name as well.

File: pkg/emptydisk/emptydisk.go [hunk 8]
```diff
@@ -42,5 +45,21 @@ func CreateTemporaryDisks(vmi *v1.VirtualMachineInstance) error {
 }
 
 func FilePathForVolumeName(volumeName string) string {
-	return path.Join(EmptyDiskBaseDir, volumeName+".qcow2")
+	return filePathForVolumeName(emptyDiskBaseDir, volumeName)
```
[the remark this answers is not in the document]
Comment (rmohr) [reply pkg/emptydisk/emptydisk.go]: Addressed.

**Review by davidvossel — APPROVED** [review]
/lgtm
/approve
